tycho-surefire/tycho-surefire-plugin/src/main/java/org/eclipse/tycho/surefire/ConcurrencyLock.java

```diff
@@ -0,0 +1,44 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Christoph Läubrich and others.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *    Christoph Läubrich - initial API and implementation
+ ******************************************************************************/
+package org.eclipse.tycho.surefire;
+
+import java.util.concurrent.Semaphore;
+
+public class ConcurrencyLock {
+
+    private Semaphore semaphore;
+    private int maxPermits;
+
+    public AutoCloseable aquire(int concurrencyLevel) throws InterruptedException {
+        if (concurrencyLevel < 1) {
+            return () -> {
+                //nothing to do
+            };
+        }
+        int aquire = 1;
+        synchronized (this) {
+            if (semaphore == null) {
+                semaphore = new Semaphore(concurrencyLevel);
```

Comment:
the semaphore is initialized with the first value of concurrencyLevel.

This all only works correctly when all projects in the reactor share the same value for concurrencyLevel, i.e. is typically configured top-level.

Would it make sense to add a consistency check (remember this first value of concurrencyLevel, compare the current one with the initial one, bail out if mismatching)?

Reply:
Ah, never mind. I didn't fully understand earlier that this is accounted for in the `else` part.

Reply:
As mentioned in the Release notes and javadoc, only the smallest number of all configured values in the reactor will be applied as it is otherwise impossible to ensure a consistent state. So one could see the configured value as an upper-bound but not of an guarantee. That's similar to requesting a parallel build on a single core processor where effectively only one CPU could be used even if a higher degree of concurrency is requested.